### ISSUE TITLE
How to make it work with legacy code

### ISSUE THREAD

**Issue by kungfooman** [Tomius/oglwrap#3]
Hey hey,

I got a bunch of old OpenGL code and I would like to use legacy OpenGL and oglwrap, but I couldn't figure out yet how to make them work together.

What I tried e.g. is to make the hidden handles public and then:

	gl::Program proggy;
	proggy.program_.handle_ = prog; // prog is GLuint
	gl::Uniform<float> u_time(proggy, "u_time");
	u_time = (float)Sys_Milliseconds();

But something like this doesn't work. Is ther any "good" way to combine old code with oglwrap?

**Comment by Tomius** [comment]
Hi,

Thanks for asking :)

I've never had to combine old OpenGL and oglwrap code, so I've never thought of this before, and currently, it is not supported.

To make oglwrap handle this use-case, I think I would replace `gl::Program` with something like
```c++
template<class HandleType>
class ProgramTemplate {
   // ...
protected:
   HandleType program_;
   // ...
};
```
And then declare the following "typedefs":

```c++
using Program = ProgramTemplate<globjects::Program>;

class LegacyProgram : public ProgramTemplate<GLuint> {
public:
   LegacyProgram(GLuint handle) { program_ = handle; }
};
```

I will think this through and make this work for all of the oglwrap classes, but I need a few days for that.

If there is a possibility to refactor the legacy code, a workaround until then is to replace the `glCreateProgram` and `glDeleteProgram` calls with a `gl::Program` prog, and then store a `GLuint handle = prog.expose()`, on which the rest of the legacy code can still work.

**Comment by Tomius** [comment]
Fixed in 6c099106ea12af0477b212776ff1fcc8ad0dbdeb

Now you can create any oglwrap object to just wrap an existing OpenGL handle using the constructor parameter, like this:
```c++
GLuint handle = glCreateProgram();
{
  gl::Program program(handle);
  // attach shaders, link and use program
}
glDeleteProgram(handle);
``` 

Edit: for an already linked program:
```c++
GLuint handle = glCreateProgram();
// attach shaders and link program in legacy code
{
  gl::Program program(handle, gl::Program::kLinkSuccessful);
  gl::Use(program); // bind the program into the OpenGL state
  // update uniforms
}
glDeleteProgram(handle);
```
